packages/contracts/src/core/permission/PermissionManager.sol

```diff
@@ -109,6 +219,210 @@ abstract contract PermissionManager is Initializable {
+    /// @param _flags The flags as uint256 the permission owner wants to remove from this specific delegatee.
+    function undelegatePermission(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _delegatee,
+        uint256 _flags
+    ) public {
+        if (_flags == 0) {
+            revert FlagCanNotBeZero();
+        }
+
+        bytes32 permHash = permissionHash(_where, _permissionIdOrSelector);
+        Permission storage permission = permissions[permHash];
+
+        if (!_checkFlags(permission.owners[msg.sender], _flags)) {
+            revert UnauthorizedOwner(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint256 currentFlags = permission.delegations[_delegatee];
+        if (!_checkFlags(currentFlags, _flags)) {
+            revert InvalidFlagsForRemovalPassed(currentFlags, _flags);
+        }
+
+        uint256 newFlags = currentFlags ^ _flags;
+        permission.delegations[_delegatee] = newFlags;
+
+        emit PermissionUndelegated(_where, _permissionIdOrSelector, _delegatee, newFlags);
+    }
+
+    /// @notice Function to add a new owner to a permission.
+    /// @param _where The address of the target contract for which `_who` receives permission.
+    /// @param _permissionIdOrSelector The permission hash or function selector used for this permission.
+    /// @param _owner The new manager for this permission.
+    /// @param _flags The flags as uint256 to restrict what this specifc owner actually can do. (only revoke? only grant? both?)
+    function addOwner(
+        address _where,
+        bytes32 _permissionIdOrSelector,
+        address _owner,
+        uint256 _flags
+    ) external {
+        if (_flags == 0) {
+            revert FlagCanNotBeZero();
+        }
+
+        if (_owner == address(0) || _where == address(0)) {
+            revert ZeroAddress();
+        }
+
+        Permission storage permission = permissions[
+            permissionHash(_where, _permissionIdOrSelector)
+        ];
+
+        if (_isPermissionFrozen(permission)) {
+            revert PermissionFrozen(_where, _permissionIdOrSelector);
+        }
+
+        if (!_checkFlags(permission.owners[msg.sender], _flags)) {
+            revert UnauthorizedOwner(msg.sender, permission.owners[msg.sender], _flags);
+        }
+
+        uint256 currentFlags = permission.owners[_owner];
+
+        if (_owner != address(1)) {
+            if (
+                _checkFlags(_flags, GRANT_OWNER_FLAG) &&
+                !_checkFlags(currentFlags, GRANT_OWNER_FLAG)
+            ) {
+                permission.grantCounter++;
+            }
+
+            if (
+                _checkFlags(_flags, REVOKE_OWNER_FLAG) &&
+                !_checkFlags(currentFlags, REVOKE_OWNER_FLAG)
+            ) {
+                permission.revokeCounter++;
+            }
+        }
+
+        permission.owners[_owner] = currentFlags | _flags; // Update owner permission
```

Comment:
I would consider checking if `currentFlags == _flags` to avoid not needed operations in case they are the same.